### PR TITLE
[mlir] Make fold result type check more verbose

### PR DESCRIPTION
**Pull request by kuhar**
Print the op and its types when the fold type check fails. This is to speed up debuging as it should be trivial to map the offending op to its folder based on the op name.

### CODE REVIEW

**Review by hanhanW — APPROVED** [review]

**Review by Groverkss — APPROVED** [review]
LGTM. This change helped a lot in debugging downstream folders.

File: mlir/lib/IR/Operation.cpp [hunk 1]
```diff
@@ -611,11 +612,19 @@ void Operation::setSuccessor(Block *block, unsigned index) {
...
+  if (results.empty())
+    return;
+
+  for (auto [ofr, opResult] : llvm::zip_equal(results, op->getResults())) {
+    if (auto value = dyn_cast<Value>(ofr)) {
+      if (value.getType() != opResult.getType()) {
+        llvm::errs() << "Folder produced a value of incorrect type for: " << *op
+                     << "\nOriginal type: '" << value.getType()
+                     << "'\nNew type: '" << opResult.getType() << "'\n";
+        assert(false && "incorrect fold result type");
```
Comment (matthias-springer) [inline mlir/lib/IR/Operation.cpp:624]: nit: You can use `llvm_unreachable("incorrect ...");`

**Review by matthias-springer — APPROVED** [review]

File: mlir/lib/IR/Operation.cpp [hunk 2]
```diff
@@ -611,11 +612,19 @@ void Operation::setSuccessor(Block *block, unsigned index) {
...
-      if (auto value = ofr.dyn_cast<Value>())
-        assert(value.getType() == opResult.getType() &&
-               "folder produced value of incorrect type");
+  if (results.empty())
+    return;
+
+  for (auto [ofr, opResult] : llvm::zip_equal(results, op->getResults())) {
+    if (auto value = dyn_cast<Value>(ofr)) {
+      if (value.getType() != opResult.getType()) {
+        llvm::errs() << "Folder produced a value of incorrect type for: " << *op
```
Comment (joker-eph) [inline mlir/lib/IR/Operation.cpp:621]: op->emitOpError() ?

**Review by joker-eph — APPROVED** [review]
Nice! I am a big proponent of making this kind of error as verbose as needed.

File: mlir/lib/IR/Operation.cpp (hunk 1 above)
Comment (kuhar) [reply mlir/lib/IR/Operation.cpp:624]: I do want the error message to be printed though and that code to be reachable. Probably doesn't matter as long as this code if wrapped in NDEBUG, but I remember there was a long discussion on this at some point that made me hesitant to use unreachable in cases like this one: https://discourse.llvm.org/t/llvm-unreachable-is-widely-misused/60587

File: mlir/lib/IR/Operation.cpp (hunk 2 above)
Comment (kuhar) [reply mlir/lib/IR/Operation.cpp]: thanks!

File: mlir/lib/IR/Operation.cpp (hunk 1 above)
Comment (joker-eph) [reply mlir/lib/IR/Operation.cpp:624]: >  I remember there was a long discussion on this at some point that made me hesitant to use unreachable in cases like this one: https://discourse.llvm.org/t/llvm-unreachable-is-widely-misused/60587

In LLVM, I don't see any case for `assert(false`.
I believe we should always use `llvm_unreachable` instead. We already have means to control the behavior:
https://github.com/llvm/llvm-project/commit/6316129e066e
(this patch is at the end of the thread you link to).

**Pull request merged**